src/base/RegexPattern.h

```diff
@@ -26,33 +24,28 @@ class RegexPattern
 
 public:
     RegexPattern() = delete;
-    RegexPattern(const SBuf &aPattern, int aFlags);
-    ~RegexPattern();
-
     RegexPattern(RegexPattern &&) = delete; // no copying of any kind
+    RegexPattern(const SBuf &aPattern, std::regex::flag_type);
 
     /// whether the regex differentiates letter case
-    bool caseSensitive() const { return !(flags & REG_ICASE); }
+    bool caseSensitive() const { return !(regex.flags() & std::regex::icase); }
 
     /// whether this is an "any single character" regex (".")
     bool isDot() const { return pattern.length() == 1 && pattern[0] == '.'; }
 
-    bool match(const char *str) const {return regexec(&regex,str,0,nullptr,0)==0;}
+    bool match(const char *) const;
```

Comment:
Please also add an SBuf matching method and remove the corresponding XXX/c_str() from refreshCheck(). Please check for other places where a RegexPattern::match() caller is using an SBuf or a buffer/length pair. If the latter cases exist, please add a method for that case as well.

Those changes were waiting for this migration to std::regex (or another modern library), and they may offset some of the performance overheads added by this migration.

Reply:
Please note that the only API change to RegexPattern is the type used by flags. Whether the rest of the API supports SBuf (or not) is unrelated to std::regex.

Reply:
> Whether the rest of the API supports SBuf (or not) is unrelated to std::regex

It is related because std::regex, unlike regexec(3), supports matching unterminated buffers. The PR switching from regexec() to std::regex should add the corresponding optimized match() method, especially since Squid code is transitioning to unterminated buffers, and there is at least one existing XXX waiting for such a match() method. The related optimization will also help partially mitigate performance regression added by this PR. It is simple change. I am happy to do it myself if you prefer.

Reply:
It is unrelated because this API change can be implemented without `std::regex` etc. The scope of this PR is the internal implementation of `RegexPattern`. There are many possible followup PRs touching callers and API which are intentionally left out of scope here.

Reply:
> It is unrelated because this API change can be implemented without `std::regex`

This API could not be implemented earlier. Or, to be more precise, it should not have been implemented earlier (because it would hide potentially significant performance overheads from the caller). The migration to `std::regex` is what makes this API possible.

> The scope of this PR is the internal implementation of RegexPattern.

The scope of this PR is what we decide it is. There are many reasonable ways to define that scope, and adding this API would not make the scope of this PR worse. I think this API improvement should be added here (for all the reasons already documented and more), but I give up.

Reply:
PR 1299 does this.